### PR TITLE
Fix: legacy backend state reading and GC 

### PR DESCRIPTION
**Pull request by chivalryq**
For now if restart controller with `--controller-namespace`. The configuration state will become `GeneratingTerraformOutputs` and stuck. This is because controller can't got the legacy backend. 

Fix that `Configuration.Spec.Backend` will be overwritten if set `--controller-namespace`. Instead, now controller 
 only overwrite `Configuration.Spec.Backend`  when no backend specified.

**How is this tested**

Add both unittest and e2e test

Signed-off-by: Qiaozp <qiaozhongpei.qzp@alibaba-inc.com>

### CODE REVIEW

**Review by Somefive — APPROVED** [review]

File: controllers/configuration/backend/kubernetes.go [hunk 1]
```diff
@@ -98,12 +111,19 @@ terraform {
 func (k K8SBackend) secretName() string {
 	return fmt.Sprintf(TFBackendSecret, terraformWorkspace, k.SecretSuffix)
 }
+func (k K8SBackend) legacySecretName() string {
```
Comment (loheagn) [inline controllers/configuration/backend/kubernetes.go:114]:
```suggestion

func (k K8SBackend) legacySecretName() string {
```

**Review by loheagn — COMMENTED** [review]
If a user applies a Configuration without `spec.Backend`, then the `terraform apply` job will use the default kubernetes backend whose secret is `{Namespace: ${env.TERRAFORM_BACKEND_NAMESPACE} || ${configuration.Namespace}, Name: tfstate-default-${confguration.Name}}`.

Then, the user restarts the controller with `--controller-namespace`. In this case, the `terraform apply` job will use the default kubernetes backend whose secret is `{Namespace: ${env.TERRAFORM_BACKEND_NAMESPACE} || ${configuration.Namespace}, Name: tfstate-default-${confguration.UID}}`. 

The two backend secret are in the same namespace but have different names. Will this cause an error?

**Comment by loheagn** [comment]
Shoule we add some e2e tests for the "restart controller with `--controller-namespace`" cases?

**Comment by chivalryq** [comment]
> If a user applies a Configuration without `spec.Backend`, then the `terraform apply` job will use the default kubernetes backend whose secret is `{Namespace: ${env.TERRAFORM_BACKEND_NAMESPACE} || ${configuration.Namespace}, Name: tfstate-default-${confguration.Name}}`.
> 
> Then, the user restarts the controller with `--controller-namespace`. In this case, the `terraform apply` job will use the default kubernetes backend whose secret is `{Namespace: ${env.TERRAFORM_BACKEND_NAMESPACE} || ${configuration.Namespace}, Name: tfstate-default-${confguration.UID}}`.
> 
> The two backend secrets are in the same namespace but have different names. Will this cause an error?

There are duplications.  But it's OK. 
1. New secret won't be used. It is created for simplicity (No need for extra logic). 
2. They will be both deleted when calling  `backend.Cleanup`.

**Pull request merged**